base/src/main/scala/data/ProCompose.scala

```diff
@@ -0,0 +1,54 @@
+package scalaz
+package data 
+
+object ProCompose {
+  import typeclass.{Profunctor, Strong, Category, StrongClass, CategoryClass}
+  import typeclass.Strong._
+  import typeclass.Profunctor._
+
+  trait ProCompose[P[_, _], A, B] { } 
+
+  case class Pure[P[_, _], A, B](pab: A => B) extends ProCompose[P, A, B]
+
+  case class Composed[P[_, _], A, X, B](pab: P[A, X], more: ProCompose[P, X, B]) extends ProCompose[P, A, B]
```

Comment:
Try to implement it yourself; for example, use `Kleisli` over `F`. You don't know enough from the definition of `ProCompose` to actually apply the function to the value; so you get stuck implementing it in a tail-recursive way. If you do it the obvious way using `Kleisli.compose` it's not stack-safe.